### PR TITLE
[opinfo] nn.functional.unfold

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: facebookresearch/functorch#78

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -2605,6 +2605,21 @@ def sample_unsqueeze(op_info, device, dtype, requires_grad, **kwargs):
     return samples
 
 
+def sample_inputs_nn_unfold(op_info, device, dtype, requires_grad, **kwargs):
+    shapes = ((1, 1, 5, 5), (0, 1, 5, 5))
+    kernel_sizes = ((3, 3), (2, 2))
+    dilations = (1, 2, (1, 2))
+    paddings = (0, 1, (1, 1))
+    strides = (1, 2, (1, 2))
```
Comment (krshrimali) [inline torch/testing/_internal/common_methods_invocations.py:2613]: These are definitely interesting cases, but here are some more cases that we can cover for completeness:

* More than one batch size for input shape (`shapes`) - currently we only have a single batch size tested.
* From the documentation: `stride, padding, dilation` are optional arguments, so checking just input shape and `kernel_size` will be a good addition, I believe.
* Would you please add an integer to `kernel_sizes`? Since that's also permitted to such functions.

File: test/test_nn.py [hunk 2]
```diff
@@ -16843,16 +16843,6 @@ def test_softshrink_negative(self, device):
...
-    def test_unfold(self, device):
-        def func(x):
-            return F.unfold(x, kernel_size=(3, 3))
-        seeds = (13, 256, 811, 43, 7)
-        for sd in seeds:
-            torch.manual_seed(sd)
-            x = torch.randn(1, 1, 5, 5, device=device, requires_grad=True)
-            gradcheck(func, [x])
-            gradgradcheck(func, [x])
-
```
Comment (krshrimali) [inline test/test_nn.py:16855]: Nice attention here, thanks @kshitij12345!

**Review by krshrimali — CHANGES_REQUESTED** [review]
This looks great, thanks @kshitij12345! I added a comment on how we can a few more cases for completeness.

It's interesting though, that we have an OpInfo for `unfold` already (wasn't aware of this) but I'm not sure we need to keep it (or if it's any different) since it doesn't cover `dilation, padding, stride` args:

https://github.com/pytorch/pytorch/blob/773a8eede437de972311b21824694e4fb120f6a2/torch/testing/_internal/common_methods_invocations.py#L7687-L7694

Thanks!

**Comment by kshitij12345** [comment]
torch.unfold is actually different from `nn.functional.unfold` (function implementation as well as API). That is why a new entry.

https://github.com/pytorch/pytorch/blob/773a8eede437de972311b21824694e4fb120f6a2/aten/src/ATen/native/native_functions.yaml#L7356

https://github.com/pytorch/pytorch/blob/773a8eede437de972311b21824694e4fb120f6a2/torch/nn/functional.py#L4481

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:2613]: Do we want to test for the just passing the kernel_sizes?
https://github.com/pytorch/pytorch/blob/773a8eede437de972311b21824694e4fb120f6a2/torch/nn/functional.py#L4481

We have covered the default values already for `dilations, paddings and strides`.

Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:2613]: > Do we want to test for the just passing the kernel_sizes?

I think yes, something like: `torch.nn.unfold(input_shape, kernel_size)` will be a good test case here.

> We have covered the default values already for dilations, paddings and strides.

Yes! Since they are optional, the above test case will cover the optional property of these args.

**Review by krshrimali — APPROVED** [review]
LGTM, thanks @kshitij12345!

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -2605,6 +2605,26 @@ def sample_unsqueeze(op_info, device, dtype, requires_grad, **kwargs):
     return samples
 
 
+def sample_inputs_nn_unfold(op_info, device, dtype, requires_grad, **kwargs):
+    shapes = ((0, 1, 5, 5), (1, 1, 5, 5), (2, 1, 5, 5))
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:2609]: We should really test a case with more than one channel (e.g. the second dimension). Could we change the last shape into (2, 3, 5, 5)?

**Review by zou3519 — COMMENTED** [review]
One really minor comment about the inputs shape. But I'm also happy to merge this as-is and add the shape in a follow-up, please let me know how you'd like to proceed

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:2609]: Sure that makes sense. Will do that!

**Review by zou3519 — APPROVED** [review]

**Comment by zou3519** [comment]
I wonder if we should partition OpInfos across multiple files to avoid merge conflicts. Right now I can only really merge one of these PRs a day (two sometimes if I'm lucky) because I need to rebase internally, ship it, and then wait for the change to reach viable/strict.

**Comment by kshitij12345** [comment]
This is being pursued in https://github.com/pytorch/pytorch/pull/59871

(Though targeting a particular kind of OpInfo like nn.functional will still lead to conflicts)
